### PR TITLE
Add an implementation of the GProximal interface for TensorFlow model(s)

### LINKED ISSUE #299: add a new `g_proximal` operator to the new class using a TensorFlow model.
The operator `g_proximal` is passed as a lambda function that takes three arguments (output, thershold, input). The learned model will only need (output, input) but we can ignore the threshold parameter. Note that on [line 234 of ForwardBackward.h](https://github.com/astro-informatics/sopt/blob/development/cpp/sopt/forward_backward.h#L234) some operations are done on the arguments that should *not* be done on the learned model.

### LINKED ISSUE #331: Use standard Forward-Backward with TF model
The algorithm used in sopt in https://github.com/astro-informatics/sopt/blob/8b7f5ee77d42bfc40f4ff9d2a2e33771b8b3322c/cpp/sopt/forward_backward.h#L253 is the FISTA algorithm decribed by algorithm 2 of [Matt's paper](https://github.com/astro-informatics/sopt/files/10468319/Learned_variational_algorithms.pdf). The python code [Optimus Primal](https://github.com/astro-informatics/Optimus-Primal/blob/2706e934f61c9406570705cee42cefc5805dec1d/optimusprimal/forward_backward.py#L51) uses the standard Forward-Backward algorithm 1. We should use the same algorithm in #317

### LINKED ISSUE #339: Move Steps of Forward Backward Algorithm out of Function Arguments
To make it easier to understand the steps being made. Related to #330 and #331 

https://github.com/astro-informatics/sopt/blob/e81d1d20af0c1f7ba5b4a1c672be3b22fa57ab4f/cpp/sopt/forward_backward.h#L234

https://github.com/astro-informatics/sopt/blob/e81d1d20af0c1f7ba5b4a1c672be3b22fa57ab4f/cpp/sopt/forward_backward.h#L256

### PR DESCRIPTION
**Pull request by tkoskela**
Closes #299 
Closes #331 

TODO list
- [x] Add calls to pre and post process the image data into the right format for the model
- [x] Add tf_g_proximal.h to CMakeLists.txt
- [x] Check that it compiles
- [x] Create a test for it -- inpainting?
- [x] Instantiate cppflow::model in constructor
- [x] Clean up image conversions between t_Vector and cppflow::tensor
- [x] Figure out why solution does not converge
- [x] #331

### CODE REVIEW

**Comment by SJaffa** [comment]
CppFlow docs on finding the model call signature is in the note at the end of this section: https://serizba.github.io/cppflow/quickstart.html#complex-model-call-multi-input-output-models

**Review by CosmoMatt — APPROVED** [review]

**Comment by tkoskela** [comment]
In this PR, we are still always using the FISTA algorithm, do you think that is OK or do we still need to do #331? We would need to parametrize how `t` is updated in https://github.com/astro-informatics/sopt/blob/8b49e78cbcb824db084392f02c4f2dd6b5cf328d/cpp/sopt/forward_backward.h#L255

File: cpp/sopt/tf_g_proximal.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+#include "sopt/exception.h"
+#include "sopt/forward_backward.h"
+#include "sopt/linear_transform.h"
+#include "sopt/logging.h"
+#include "sopt/proximal.h"
+#include "sopt/relative_variation.h"
+#include "sopt/types.h"
+#include "sopt/g_proximal.h"
+
+#include <cppflow/cppflow.h>
```
Comment (SJaffa) [inline cpp/sopt/tf_g_proximal.h]:
```suggestion
#include "cppflow/cppflow.h"
```

File: cpp/tests/tf_inpainting.cc [hunk 2]
```diff
@@ -0,0 +1,86 @@
+#include <algorithm>
+#include <exception>
+#include <functional>
+#include <iostream>
+#include <random>
+#include <vector>
+#include <ctime>
+#include <catch.hpp>
+
+#include <sopt/imaging_forward_backward.h>
```
Comment (SJaffa) [inline cpp/tests/tf_inpainting.cc]:
```suggestion
#include "sopt/imaging_forward_backward.h"
```

File: cpp/tests/tf_inpainting.cc [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  sopt::t_real const gamma = 18;
+  sopt::t_real const beta = sigma * sigma * 0.5;
+
+  auto fb = sopt::algorithm::ImagingForwardBackward<Scalar>(y);
+  fb.itermax(500)
+    .beta(beta)    // stepsize
+    .sigma(sigma)  // sigma
+    .gamma(gamma)  // regularisation paramater
+    .fista(false)  // switch for FISTA algorithm, don't use with learned TF model
```
Comment (SJaffa) [inline cpp/tests/tf_inpainting.cc]:
```suggestion
    .fista(false)  // switch to use FISTA algorithm in Forward Backward algorithm, should be false if using learned TF model
```

File: cpp/sopt/tf_g_proximal.h [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  typedef typename FB::t_LinearTransform t_LinearTransform;
+
+  // The constructor constructs a cppflow model object from a saved model saved
+  // to the file filename
+  TFGProximal(std::string path)
+    : model_(path),
+      square_image_(true),
+      Psi_(linear_transform_identity<Scalar>())
+  {
+  //   for ( auto str : model_.get_operations() ) {
```
Comment (SJaffa) [inline cpp/sopt/tf_g_proximal.h]: Should we remove this print-debugging or keep it for later use?

**Pull request merged**